src/openassetio-core/src/managerApi/ManagerInterface.cpp

```diff
@@ -48,6 +48,18 @@ ManagerStateBasePtr ManagerInterface::stateFromPersistenceToken(
 EntityReference ManagerInterface::createEntityReference(Str entityReferenceString) const {
   return EntityReference(std::move(entityReferenceString));
 }
+
+void ManagerInterface::getWithRelationship(
+    [[maybe_unused]] const TraitsDataPtr& relationshipTraitsData,
+    const EntityReferences& entityReferences, [[maybe_unused]] const ContextConstPtr& context,
+    [[maybe_unused]] const HostSessionPtr& hostSession,
+    const ManagerInterface::RelationshipSuccessCallback& successCallback,
+    [[maybe_unused]] const ManagerInterface::BatchElementErrorCallback& errorCallback,
+    [[maybe_unused]] const trait::TraitSet& resultTraitSet) {
+  for (std::size_t idx = 0; idx < entityReferences.size(); ++idx) {
```

Comment:
Good shout. Some interesting [reading](https://stackoverflow.com/a/58553484/535103) and [experimentation](https://godbolt.org/z/b96Wcfc1f) led me to conclude:

In terms of macro-optimisation: `.size()` is almost definitely inlined and for `std::vector` is just a direct memory access. If we could swap out the container type (e.g. if the function was templated on container type) then it would be more of a concern.

In terms of micro-optimisation: in our case the compiler cannot _prove_ that `.size()` always returns the same value at each iteration, because we call out to another function, which could conceivably modify the vector.  So the compiler must _always_ fetch the latest value of `.size()` from memory, rather than a CPU register.

Good learnings.  I'll do the change.

Reply:
Interesting. I did some [quick-bench](https://quick-bench.com/q/CziHGLS6Weuo1pk6sVejRRzbJ1k) benchmarks. Over 1000 elements. Clang 15.0
| InlineSizeCheck | HoistedSizeCheck | Normalized Form |
| ----------- | ----------- |-----------|
| 252.4 | 250.8 | 1 : 1.0063 |
| 251.3 | 249.9 | 1 : 1.0056 |
| 247.5 | 247.7 | 1 : 0.9992 |
| 250.8 | 251.6 | 1 : 0.9968 |
| 250.9 | 248.4 | 1 : 1.0101 |

As data ... this is worthless! :medal_sports: Smells like the difference being small enough that random variance in testing it overrides any real information.

Reply:
I [tweaked quick-bench](https://quick-bench.com/q/PJuXiiWIUU6oS_gPxnKaR3ItMDg) code to be a bit more representative, and do see a small performance difference:

![image](https://github.com/OpenAssetIO/OpenAssetIO/assets/7827094/e3eeaee4-febb-41d1-bc33-9e2ce72e1283)

Reply:
Explicitly setting the function to not be optimised smells a tad unfair to me a bit, but certainly this is quite interesting. I might quibble, especially in the presence of LTO, but there's certainly an argument.

Reply:
Micro-optimised in bc2b0f5c and 9562c994.

Tempted to drop though favouring slightly better readability over imperceptibly better performance.  The work done in the the loop will dominate any real-world timings _by far_.

I guess the key thing is - will contributors/adopters of OpenAssetIO feel more confident in a project that prioritises micro-optimisation vs. maximising readability?  I suspect, this day and age, readability has higher weighting.

Reply:
After asking around internally, there is no clear consensus.  So I'll make an executive decision to leave as is (i.e. with the micro-optimisation in place). It's slightly more future-proof against swapping out the container, and shows any inquisitive engineers that we're considerate of C++ quirks.